File: sdks/swift/Tests/EdgeQuakeSDKTests/UnitTest.swift
import Foundation
import XCTest

@testable import EdgeQuakeSDK

// MARK: - MockURLProtocol

/// Mock URL protocol that returns predefined responses without network calls.
/// WHY: Enables stateless unit testing of all service methods.
final class MockURLProtocol: URLProtocol {
    static var responseData: Data = "{}".data(using: .utf8)!
    static var responseStatusCode: Int = 200
    static var requestHistory: [(method: String, url: String, body: Data?)] = []

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }

    override func startLoading() {
        let method = request.httpMethod ?? "GET"
        let url = request.url?.absoluteString ?? ""
        // WHY: URLProtocol strips httpBody; read from httpBodyStream instead
        var body = request.httpBody
        if body == nil, let stream = request.httpBodyStream {
            stream.open()
            var data = Data()
            let bufferSize = 4096
            let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
            defer { buffer.deallocate() }
            while stream.hasBytesAvailable {
                let read = stream.read(buffer, maxLength: bufferSize)
                if read > 0 {
                    data.append(buffer, count: read)
                } else {
                    break
                }
            }
            stream.close()
            body = data.isEmpty ? nil : data
        }
        MockURLProtocol.requestHistory.append((method: method, url: url, body: body))

        let response = HTTPURLResponse(
            url: request.url!, statusCode: MockURLProtocol.responseStatusCode,
            httpVersion: "HTTP/1.1", headerFields: ["Content-Type": "application/json"]
        )!
        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
        client?.urlProtocol(self, didLoad: MockURLProtocol.responseData)
        client?.urlProtocolDidFinishLoading(self)
    }

    override func stopLoading() {}

    static func reset(json: String = "{}", status: Int = 200) {
        responseData = json.data(using: .utf8)!
        responseStatusCode = status
        requestHistory = []
    }

    static var lastRequest: (method: String, url: String, body: Data?)? {
        requestHistory.last
    }
}

// MARK: - Test Helpers

func mockHelper(json: String = "{}", status: Int = 200) -> HttpHelper {
    MockURLProtocol.reset(json: json, status: status)
    let config = URLSessionConfiguration.ephemeral
    config.protocolClasses = [MockURLProtocol.self]
    let session = URLSession(configuration: config)
    return HttpHelper(config: EdgeQuakeConfig(), session: session)
}

// MARK: - Config Tests

final class ConfigTest: XCTestCase {
    func testDefaults() {
        let c = EdgeQuakeConfig()
        XCTAssertEqual(c.baseUrl, "http://localhost:8080")
        XCTAssertNil(c.apiKey)
        XCTAssertNil(c.tenantId)
        XCTAssertNil(c.userId)
        XCTAssertNil(c.workspaceId)
        XCTAssertEqual(c.timeoutSeconds, 30)
    }

    func testCustomValues() {
        let c = EdgeQuakeConfig(
            baseUrl: "https://api.example.com", apiKey: "sk-test",
            tenantId: "t-1", userId: "u-1", workspaceId: "ws-1", timeoutSeconds: 120
        )
        XCTAssertEqual(c.baseUrl, "https://api.example.com")
        XCTAssertEqual(c.apiKey, "sk-test")
        XCTAssertEqual(c.tenantId, "t-1")
        XCTAssertEqual(c.userId, "u-1")
        XCTAssertEqual(c.workspaceId, "ws-1")
        XCTAssertEqual(c.timeoutSeconds, 120)
    }
}

// MARK: - Error Tests

final class ErrorTest: XCTestCase {
    func testProperties() {
        let err = EdgeQuakeError(message: "bad request", statusCode: 400, responseBody: "{}")
        XCTAssertEqual(err.message, "bad request")
        XCTAssertEqual(err.statusCode, 400)
        XCTAssertEqual(err.responseBody, "{}")
        XCTAssertEqual(err.errorDescription, "bad request")
    }

    func testIsError() {
        let err = EdgeQuakeError(message: "test")
        XCTAssertTrue(err is Error)
    }

    func testDefaults() {
        let err = EdgeQuakeError(message: "test")
        XCTAssertEqual(err.statusCode, 0)
        XCTAssertNil(err.responseBody)
    }
}

// MARK: - Client Tests

final class ClientTest: XCTestCase {
    func testInitializesAllServices() {
        let client = EdgeQuakeClient()
        XCTAssertNotNil(client.health)
        XCTAssertNotNil(client.documents)
        XCTAssertNotNil(client.entities)
        XCTAssertNotNil(client.relationships)
        XCTAssertNotNil(client.graph)
        XCTAssertNotNil(client.query)
        XCTAssertNotNil(client.chat)
        XCTAssertNotNil(client.tenants)
        XCTAssertNotNil(client.users)
        XCTAssertNotNil(client.apiKeys)
        XCTAssertNotNil(client.tasks)
        XCTAssertNotNil(client.pipeline)
        XCTAssertNotNil(client.models)
        XCTAssertNotNil(client.costs)
    }
}

// MARK: - Health Tests

final class HealthServiceTest: XCTestCase {
    func testCheck() async throws {
        let http = mockHelper(json: #"{"status":"healthy","version":"0.1.0"}"#)
        let svc = HealthService(http)
        let result = try await svc.check()
        XCTAssertEqual(result.status, "healthy")
        XCTAssertEqual(result.version, "0.1.0")
        XCTAssertEqual(MockURLProtocol.lastRequest?.method, "GET")
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("/health"))
    }

    func testCheckError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = HealthService(http)
        do {
            _ = try await svc.check()
            XCTFail("Expected error")
        } catch {
            if let err = error as? EdgeQuakeError {
                XCTAssertEqual(err.statusCode, 500)
            }
        }
    }
}

// MARK: - Document Tests

final class DocumentServiceTest: XCTestCase {
    func testList() async throws {
        let http = mockHelper(json: #"{"documents":[{"id":"d1","title":"Test"}],"total":1}"#)
        let svc = DocumentService(http)
        let result = try await svc.list()
        XCTAssertEqual(result.documents?.count, 1)
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("page=1"))
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("page_size=20"))
    }

    func testListPagination() async throws {
        let http = mockHelper(json: #"{"documents":[]}"#)
        let svc = DocumentService(http)
        _ = try await svc.list(page: 3, pageSize: 50)
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("page=3"))
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("page_size=50"))
    }

    func testGet() async throws {
        let http = mockHelper(json: #"{"id":"d1","title":"Test"}"#)
        let svc = DocumentService(http)
        let result = try await svc.get(id: "d1")
        XCTAssertEqual(result.id, "d1")
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("/api/v1/documents/d1"))
    }

    func testUploadText() async throws {
        let http = mockHelper(json: #"{"document_id":"d2","status":"processing"}"#)
        let svc = DocumentService(http)
        let result = try await svc.uploadText(title: "My Title", content: "Hello World")
        XCTAssertEqual(result.documentId, "d2")
        XCTAssertEqual(MockURLProtocol.lastRequest?.method, "POST")
    }

    func testDelete() async throws {
        let http = mockHelper(json: #"{"status":"deleted"}"#)
        let svc = DocumentService(http)
        _ = try await svc.delete(id: "d1")
        XCTAssertEqual(MockURLProtocol.lastRequest?.method, "DELETE")
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("/api/v1/documents/d1"))
    }

    func testListError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = DocumentService(http)
        do {
            _ = try await svc.list()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Entity Tests

final class EntityServiceTest: XCTestCase {
    func testList() async throws {
        let http = mockHelper(json: #"{"items":[{"entity_name":"ALICE"}],"total":1}"#)
        let svc = EntityService(http)
        let result = try await svc.list()
        XCTAssertEqual(result.items?.count, 1)
    }

    func testGet() async throws {
        let http = mockHelper(json: #"{"entity":{"entity_name":"ALICE"}}"#)
        let svc = EntityService(http)
        _ = try await svc.get(name: "ALICE")
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("/api/v1/graph/entities/ALICE"))
    }

    func testCreate() async throws {
        let http = mockHelper(json: #"{"status":"success"}"#)
        let svc = EntityService(http)
        let req = CreateEntityRequest(
            entityName: "BOB", entityType: "person", description: "A person", sourceId: "src-1")
        let result = try await svc.create(req)
        XCTAssertEqual(result.status, "success")
        XCTAssertEqual(MockURLProtocol.lastRequest?.method, "POST")
    }

    func testDelete() async throws {
        let http = mockHelper(json: #"{"status":"deleted"}"#)
        let svc = EntityService(http)
        _ = try await svc.delete(name: "BOB")
        XCTAssertEqual(MockURLProtocol.lastRequest?.method, "DELETE")
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("confirm=true"))
    }

    func testExists() async throws {
        let http = mockHelper(json: #"{"exists":true}"#)
        let svc = EntityService(http)
        let result = try await svc.exists(name: "ALICE")
        XCTAssertEqual(result.exists, true)
    }

    func testListError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = EntityService(http)
        do {
            _ = try await svc.list()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Relationship Tests

final class RelationshipServiceTest: XCTestCase {
    func testList() async throws {
        let http = mockHelper(json: #"{"items":[{"source":"A","target":"B"}],"total":1}"#)
        let svc = RelationshipService(http)
        let result = try await svc.list()
        XCTAssertEqual(result.items?.count, 1)
    }

    func testListPagination() async throws {
        let http = mockHelper(json: #"{"items":[]}"#)
        let svc = RelationshipService(http)
        _ = try await svc.list(page: 2, pageSize: 10)
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("page=2"))
    }

    func testListError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = RelationshipService(http)
        do {
            _ = try await svc.list()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Graph Tests

final class GraphServiceTest: XCTestCase {
    func testGet() async throws {
        let http = mockHelper(json: #"{"nodes":[],"edges":[]}"#)
        let svc = GraphService(http)
        let result = try await svc.get()
        XCTAssertNotNil(result.nodes)
    }

    func testSearch() async throws {
        let http = mockHelper(json: #"{"nodes":[{"id":"n1"}]}"#)
        let svc = GraphService(http)
        let result = try await svc.search(query: "Alice")
        XCTAssertEqual(result.nodes?.count, 1)
        XCTAssertTrue(MockURLProtocol.lastRequest!.url.contains("q=Alice"))
    }

    func testGetError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = GraphService(http)
        do {
            _ = try await svc.get()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Query Tests

final class QueryServiceTest: XCTestCase {
    func testExecute() async throws {
        let http = mockHelper(json: #"{"answer":"42","sources":[]}"#)
        let svc = QueryService(http)
        let result = try await svc.execute(query: "meaning of life")
        XCTAssertEqual(result.answer, "42")
        XCTAssertEqual(MockURLProtocol.lastRequest?.method, "POST")
    }

    func testExecuteWithMode() async throws {
        let http = mockHelper(json: #"{"answer":"yes"}"#)
        let svc = QueryService(http)
        _ = try await svc.execute(query: "test", mode: "local")
        guard let body = MockURLProtocol.lastRequest?.body,
            let str = String(data: body, encoding: .utf8)
        else {
            XCTFail("No body")
            return
        }
        XCTAssertTrue(str.contains("local"))
    }

    func testExecuteError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = QueryService(http)
        do {
            _ = try await svc.execute(query: "test")
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Chat Tests

final class ChatServiceTest: XCTestCase {
    func testCompletions() async throws {
        let http = mockHelper(json: #"{"content":"Hello!"}"#)
        let svc = ChatService(http)
        let req = ChatCompletionRequest(message: "Hi")
        let result = try await svc.completions(req)
        XCTAssertEqual(result.content, "Hello!")
        XCTAssertEqual(MockURLProtocol.lastRequest?.method, "POST")
    }

    func testCompletionsError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = ChatService(http)
        let req = ChatCompletionRequest(message: "test")
        do {
            _ = try await svc.completions(req)
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Tenant Tests

final class TenantServiceTest: XCTestCase {
    func testList() async throws {
        let http = mockHelper(json: #"{"items":[{"id":"t1"}]}"#)
        let svc = TenantService(http)
        let result = try await svc.list()
        XCTAssertEqual(result.items?.count, 1)
    }

    func testListError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = TenantService(http)
        do {
            _ = try await svc.list()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - User Tests

final class UserServiceTest: XCTestCase {
    func testList() async throws {
        let http = mockHelper(json: #"{"users":[{"id":"u1"}]}"#)
        let svc = UserService(http)
        let result = try await svc.list()
        XCTAssertEqual(result.users?.count, 1)
    }

    func testListError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = UserService(http)
        do {
            _ = try await svc.list()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - API Key Tests

final class ApiKeyServiceTest: XCTestCase {
    func testList() async throws {
        let http = mockHelper(json: #"{"keys":[{"id":"ak-1"}]}"#)
        let svc = ApiKeyService(http)
        let result = try await svc.list()
        XCTAssertEqual(result.keys?.count, 1)
    }

    func testListError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = ApiKeyService(http)
        do {
            _ = try await svc.list()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Task Tests

final class TaskServiceTest: XCTestCase {
    func testList() async throws {
        let http = mockHelper(json: #"{"tasks":[{"track_id":"trk-1"}]}"#)
        let svc = TaskService(http)
        let result = try await svc.list()
        XCTAssertEqual(result.tasks?.count, 1)
    }

    func testListError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = TaskService(http)
        do {
            _ = try await svc.list()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Pipeline Tests

final class PipelineServiceTest: XCTestCase {
    func testStatus() async throws {
        let http = mockHelper(json: #"{"is_busy":true,"pending_tasks":5}"#)
        let svc = PipelineService(http)
        let result = try await svc.status()
        XCTAssertEqual(result.isBusy, true)
    }

    func testQueueMetrics() async throws {
        let http = mockHelper(json: #"{"pending_count":10}"#)
        let svc = PipelineService(http)
        let result = try await svc.queueMetrics()
        XCTAssertEqual(result.pendingCount, 10)
    }

    func testStatusError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = PipelineService(http)
        do {
            _ = try await svc.status()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Model Tests

final class ModelServiceTest: XCTestCase {
    func testCatalog() async throws {
        let http = mockHelper(json: #"{"providers":[{"name":"openai"}]}"#)
        let svc = ModelService(http)
        let result = try await svc.catalog()
        XCTAssertEqual(result.providers?.count, 1)
    }

    func testHealth() async throws {
        let http = mockHelper(json: #"[{"name":"ollama","enabled":true}]"#)
        let svc = ModelService(http)
        let result = try await svc.health()
        XCTAssertEqual(result.count, 1)
        XCTAssertEqual(result[0].name, "ollama")
    }

    func testProviderStatus() async throws {
        let http = mockHelper(json: #"{"provider":{"name":"ollama"}}"#)
        let svc = ModelService(http)
        let result = try await svc.providerStatus()
        XCTAssertNotNil(result.provider)
    }

    func testCatalogError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = ModelService(http)
        do {
            _ = try await svc.catalog()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Cost Tests

final class CostServiceTest: XCTestCase {
    func testSummary() async throws {
        let http = mockHelper(json: #"{"total_cost":12.5}"#)
        let svc = CostService(http)
        let result = try await svc.summary()
        XCTAssertEqual(result.totalCost, 12.5)
    }

    func testSummaryError() async {
        let http = mockHelper(json: "{}", status: 500)
        let svc = CostService(http)
        do {
            _ = try await svc.summary()
            XCTFail("Expected error")
        } catch {
            // OK
        }
    }
}

// MARK: - Mock Tests

final class MockTests: XCTestCase {
    func testTracksAllCalls() async throws {
        let http = mockHelper(json: #"{"status":"healthy"}"#)
        let svc = HealthService(http)
        _ = try await svc.check()
        _ = try await svc.check()
        XCTAssertEqual(MockURLProtocol.requestHistory.count, 2)
    }
}

File: sdks/ruby/test/unit_test.rb
# frozen_string_literal: true

require "minitest/autorun"
require_relative "mock_http_helper"

# Unit tests for the EdgeQuake Ruby SDK.
# WHY: Verify all components without making real HTTP calls.
module EdgeQuake
  class ConfigTest < Minitest::Test
    def test_defaults
      c = Config.new
      assert_equal "http://localhost:8080", c.base_url
      assert_nil c.api_key
      assert_nil c.tenant_id
      assert_nil c.user_id
      assert_nil c.workspace_id
      assert_equal 60, c.timeout
    end

    def test_custom_values
      c = Config.new(
        base_url: "https://api.example.com",
        api_key: "sk-test",
        tenant_id: "t-1",
        user_id: "u-1",
        workspace_id: "ws-1",
        timeout: 120
      )
      assert_equal "https://api.example.com", c.base_url
      assert_equal "sk-test", c.api_key
      assert_equal "t-1", c.tenant_id
      assert_equal "u-1", c.user_id
      assert_equal "ws-1", c.workspace_id
      assert_equal 120, c.timeout
    end

    def test_strips_trailing_slash
      c = Config.new(base_url: "http://localhost:8080/")
      assert_equal "http://localhost:8080", c.base_url
    end
  end

  class ApiErrorTest < Minitest::Test
    def test_message_and_properties
      err = ApiError.new("bad request", status_code: 400, response_body: '{"error":"fail"}')
      assert_equal "bad request", err.message
      assert_equal 400, err.status_code
      assert_equal '{"error":"fail"}', err.response_body
    end

    def test_is_standard_error
      err = ApiError.new("test")
      assert_kind_of StandardError, err
    end

    def test_nil_defaults
      err = ApiError.new("test")
      assert_nil err.status_code
      assert_nil err.response_body
    end
  end

  class ClientTest < Minitest::Test
    def test_initializes_all_services
      client = Client.new
      assert_instance_of HealthService, client.health
      assert_instance_of DocumentService, client.documents
      assert_instance_of EntityService, client.entities
      assert_instance_of RelationshipService, client.relationships
      assert_instance_of GraphService, client.graph
      assert_instance_of QueryService, client.query
      assert_instance_of ChatService, client.chat
      assert_instance_of TenantService, client.tenants
      assert_instance_of UserService, client.users
      assert_instance_of ApiKeyService, client.api_keys
      assert_instance_of TaskService, client.tasks
      assert_instance_of PipelineService, client.pipeline
      assert_instance_of ModelService, client.models
      assert_instance_of CostService, client.costs
    end
  end

  class HealthServiceTest < Minitest::Test
    def test_check
      mock = MockHttpHelper.new('{"status":"healthy","version":"0.1.0"}')
      svc = HealthService.new(mock)
      result = svc.check
      assert_equal "healthy", result["status"]
      assert_equal :get, mock.last_call[:method]
      assert_equal "/health", mock.last_call[:path]
    end

    def test_check_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = HealthService.new(mock)
      assert_raises(ApiError) { svc.check }
    end
  end

  class DocumentServiceTest < Minitest::Test
    def test_list
      mock = MockHttpHelper.new('{"documents":[{"id":"d1"}]}')
      svc = DocumentService.new(mock)
      result = svc.list
      assert_equal 1, result["documents"].size
      assert_includes mock.last_call[:path], "page=1"
      assert_includes mock.last_call[:path], "page_size=20"
    end

    def test_list_pagination
      mock = MockHttpHelper.new('{"documents":[]}')
      svc = DocumentService.new(mock)
      svc.list(page: 3, page_size: 50)
      assert_includes mock.last_call[:path], "page=3"
      assert_includes mock.last_call[:path], "page_size=50"
    end

    def test_get
      mock = MockHttpHelper.new('{"id":"d1","file_name":"test.pdf"}')
      svc = DocumentService.new(mock)
      result = svc.get(id: "d1")
      assert_equal "d1", result["id"]
      assert_includes mock.last_call[:path], "/api/v1/documents/d1"
    end

    def test_upload_text
      mock = MockHttpHelper.new('{"id":"d2","status":"processing"}')
      svc = DocumentService.new(mock)
      result = svc.upload_text(title: "My Title", content: "Hello World")
      assert_equal "d2", result["id"]
      assert_equal :post, mock.last_call[:method]
      assert_equal "My Title", mock.last_call[:body][:title]
    end

    def test_upload_text_custom_file_type
      mock = MockHttpHelper.new('{"id":"d3"}')
      svc = DocumentService.new(mock)
      svc.upload_text(title: "T", content: "C", file_type: "md")
      assert_equal "md", mock.last_call[:body][:file_type]
    end

    def test_delete
      mock = MockHttpHelper.new('{"status":"deleted"}')
      svc = DocumentService.new(mock)
      svc.delete(id: "d1")
      assert_equal :delete, mock.last_call[:method]
      assert_includes mock.last_call[:path], "/api/v1/documents/d1"
    end

    def test_list_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = DocumentService.new(mock)
      assert_raises(ApiError) { svc.list }
    end

    def test_get_error
      mock = MockHttpHelper.new.will_return("{}", 404)
      svc = DocumentService.new(mock)
      assert_raises(ApiError) { svc.get(id: "missing") }
    end
  end

  class EntityServiceTest < Minitest::Test
    def test_list
      mock = MockHttpHelper.new('{"items":[{"entity_name":"ALICE"}],"total":1}')
      svc = EntityService.new(mock)
      result = svc.list
      assert_equal 1, result["items"].size
      assert_equal "ALICE", result["items"][0]["entity_name"]
    end

    def test_get
      mock = MockHttpHelper.new('{"entity_name":"ALICE","entity_type":"person"}')
      svc = EntityService.new(mock)
      result = svc.get(name: "ALICE")
      assert_equal "person", result["entity_type"]
    end

    def test_create
      mock = MockHttpHelper.new('{"status":"success"}')
      svc = EntityService.new(mock)
      result = svc.create(entity_name: "BOB", entity_type: "person", description: "A person", source_id: "src-1")
      assert_equal "success", result["status"]
      assert_equal :post, mock.last_call[:method]
      assert_equal "BOB", mock.last_call[:body][:entity_name]
      assert_equal "person", mock.last_call[:body][:entity_type]
    end

    def test_delete
      mock = MockHttpHelper.new('{}')
      svc = EntityService.new(mock)
      svc.delete(name: "BOB")
      assert_equal :delete, mock.last_call[:method]
      assert_includes mock.last_call[:path], "confirm=true"
    end

    def test_exists
      mock = MockHttpHelper.new('{"exists":true}')
      svc = EntityService.new(mock)
      result = svc.exists?(name: "ALICE")
      assert_equal true, result["exists"]
      assert_includes mock.last_call[:path], "entity_name=ALICE"
    end

    def test_list_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = EntityService.new(mock)
      assert_raises(ApiError) { svc.list }
    end
  end

  class RelationshipServiceTest < Minitest::Test
    def test_list
      mock = MockHttpHelper.new('{"items":[{"source":"A","target":"B"}],"total":1}')
      svc = RelationshipService.new(mock)
      result = svc.list
      assert_equal 1, result["items"].size
    end

    def test_list_pagination
      mock = MockHttpHelper.new('{"items":[]}')
      svc = RelationshipService.new(mock)
      svc.list(page: 2, page_size: 10)
      assert_includes mock.last_call[:path], "page=2"
      assert_includes mock.last_call[:path], "page_size=10"
    end

    def test_list_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = RelationshipService.new(mock)
      assert_raises(ApiError) { svc.list }
    end
  end

  class GraphServiceTest < Minitest::Test
    def test_get
      mock = MockHttpHelper.new('{"nodes":[],"edges":[]}')
      svc = GraphService.new(mock)
      result = svc.get
      assert result.key?("nodes")
    end

    def test_search
      mock = MockHttpHelper.new('{"nodes":[{"id":"n1"}]}')
      svc = GraphService.new(mock)
      result = svc.search(query: "Alice")
      assert_equal 1, result["nodes"].size
      assert_includes mock.last_call[:path], "q=Alice"
    end

    def test_search_url_encoding
      mock = MockHttpHelper.new('{"nodes":[]}')
      svc = GraphService.new(mock)
      svc.search(query: "hello world")
      assert_includes mock.last_call[:path], "q=hello+world"
    end

    def test_get_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = GraphService.new(mock)
      assert_raises(ApiError) { svc.get }
    end

    def test_search_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = GraphService.new(mock)
      assert_raises(ApiError) { svc.search(query: "test") }
    end
  end

  class QueryServiceTest < Minitest::Test
    def test_execute
      mock = MockHttpHelper.new('{"answer":"42","sources":[]}')
      svc = QueryService.new(mock)
      result = svc.execute(query: "meaning of life")
      assert_equal "42", result["answer"]
      assert_equal :post, mock.last_call[:method]
      assert_equal "meaning of life", mock.last_call[:body][:query]
    end

    def test_execute_with_mode
      mock = MockHttpHelper.new('{"answer":"yes"}')
      svc = QueryService.new(mock)
      svc.execute(query: "test", mode: "local")
      assert_equal "local", mock.last_call[:body][:mode]
    end

    def test_execute_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = QueryService.new(mock)
      assert_raises(ApiError) { svc.execute(query: "test") }
    end
  end

  class ChatServiceTest < Minitest::Test
    def test_completions
      mock = MockHttpHelper.new('{"choices":[{"message":{"content":"Hello!"}}]}')
      svc = ChatService.new(mock)
      result = svc.completions(message: "Hi")
      assert_equal 1, result["choices"].size
      assert_equal :post, mock.last_call[:method]
    end

    def test_completions_with_options
      mock = MockHttpHelper.new('{"choices":[]}')
      svc = ChatService.new(mock)
      svc.completions(message: "hi", mode: "global", stream: true)
      body = mock.last_call[:body]
      assert_equal "hi", body[:message]
      assert_equal "global", body[:mode]
      assert_equal true, body[:stream]
    end

    def test_completions_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = ChatService.new(mock)
      assert_raises(ApiError) { svc.completions(message: "test") }
    end
  end

  class TenantServiceTest < Minitest::Test
    def test_list
      mock = MockHttpHelper.new('{"items":[{"id":"t1"}]}')
      svc = TenantService.new(mock)
      result = svc.list
      assert_equal 1, result["items"].size
    end

    def test_list_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = TenantService.new(mock)
      assert_raises(ApiError) { svc.list }
    end
  end

  class UserServiceTest < Minitest::Test
    def test_list
      mock = MockHttpHelper.new('[{"id":"u1","username":"admin"}]')
      svc = UserService.new(mock)
      result = svc.list
      assert_equal 1, result.size
    end

    def test_list_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = UserService.new(mock)
      assert_raises(ApiError) { svc.list }
    end
  end

  class ApiKeyServiceTest < Minitest::Test
    def test_list
      mock = MockHttpHelper.new('[{"id":"ak-1"}]')
      svc = ApiKeyService.new(mock)
      result = svc.list
      assert_equal 1, result.size
    end

    def test_list_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = ApiKeyService.new(mock)
      assert_raises(ApiError) { svc.list }
    end
  end

  class TaskServiceTest < Minitest::Test
    def test_list
      mock = MockHttpHelper.new('{"tasks":[{"track_id":"trk-1"}]}')
      svc = TaskService.new(mock)
      result = svc.list
      assert_equal 1, result["tasks"].size
    end

    def test_list_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = TaskService.new(mock)
      assert_raises(ApiError) { svc.list }
    end
  end

  class PipelineServiceTest < Minitest::Test
    def test_status
      mock = MockHttpHelper.new('{"is_busy":true,"pending_tasks":5}')
      svc = PipelineService.new(mock)
      result = svc.status
      assert_equal true, result["is_busy"]
    end

    def test_queue_metrics
      mock = MockHttpHelper.new('{"queue_depth":10}')
      svc = PipelineService.new(mock)
      result = svc.queue_metrics
      assert_equal 10, result["queue_depth"]
    end

    def test_status_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = PipelineService.new(mock)
      assert_raises(ApiError) { svc.status }
    end

    def test_queue_metrics_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = PipelineService.new(mock)
      assert_raises(ApiError) { svc.queue_metrics }
    end
  end

  class ModelServiceTest < Minitest::Test
    def test_catalog
      mock = MockHttpHelper.new('{"providers":[{"name":"openai"}]}')
      svc = ModelService.new(mock)
      result = svc.catalog
      assert_equal 1, result["providers"].size
    end

    def test_health
      mock = MockHttpHelper.new('{"status":"ok","models":["qwen2.5"]}')
      svc = ModelService.new(mock)
      result = svc.health
      assert_equal "ok", result["status"]
    end

    def test_provider_status
      mock = MockHttpHelper.new('{"current_provider":"ollama"}')
      svc = ModelService.new(mock)
      result = svc.provider_status
      assert_equal "ollama", result["current_provider"]
    end

    def test_catalog_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = ModelService.new(mock)
      assert_raises(ApiError) { svc.catalog }
    end

    def test_health_error
      mock = MockHttpHelper.new.will_return("{}", 502)
      svc = ModelService.new(mock)
      assert_raises(ApiError) { svc.health }
    end
  end

  class CostServiceTest < Minitest::Test
    def test_summary
      mock = MockHttpHelper.new('{"total_cost_usd":12.5}')
      svc = CostService.new(mock)
      result = svc.summary
      assert_equal 12.5, result["total_cost_usd"]
    end

    def test_summary_error
      mock = MockHttpHelper.new.will_return("{}", 500)
      svc = CostService.new(mock)
      assert_raises(ApiError) { svc.summary }
    end
  end

  class MockHttpHelperTest < Minitest::Test
    def test_tracks_all_calls
      mock = MockHttpHelper.new("{}")
      svc = HealthService.new(mock)
      svc.check
      svc.check
      assert_equal 2, mock.calls.size
    end

    def test_error_includes_status_code
      mock = MockHttpHelper.new.will_return('{"error":"not found"}', 404)
      svc = HealthService.new(mock)
      err = assert_raises(ApiError) { svc.check }
      assert_equal 404, err.status_code
    end

    def test_will_return_chaining
      mock = MockHttpHelper.new.will_return('{"a":1}', 200)
      svc = HealthService.new(mock)
      result = svc.check
      assert_equal 1, result["a"]
    end
  end
end
